src/main/java/sirius/biz/tycho/academy/OnboardingEngine.java

```diff
@@ -0,0 +1,316 @@
+     */
+    protected abstract void markOutdatedAcademyVideosAsDeleted(String academy, String syncTokenToIgnore);
+
+    /**
+     * Lists all videos of the given academy.
+     *
+     * @param academy the academy to fetch the videos for
+     * @return the videos in the given academy
+     */
+    protected abstract List<? extends AcademyVideo> queryAcademyVideos(String academy);
+
+    /**
+     * Actually pulls the metadata of the requested academy video from the database.
+     *
+     * @param videoId the id of the video to load
+     * @return the loaded video or <tt>null</tt> if the video is unknown
+     */
+    @Nullable
+    protected abstract AcademyVideo fetchAcademyVideoFromDatabase(String videoId);
+
+    /**
+     * Creates or find the onboarding video for the given owner/target and academy video.
+     *
+     * @param owner        the onboarding participant which is targeted by the video
+     * @param academyVideo the video to show to the participant
+     * @param syncToken    the token to store in the onboarding video to support updates / cleanups
+     * @return the onboarding video for the given academy video and target
+     */
+    protected abstract OnboardingVideo createOrUpdateOnboardingVideo(String owner,
+                                                                     AcademyVideo academyVideo,
+                                                                     String syncToken);
+
+    /**
+     * Marks all untouched onboarding videos as outdated.
+     *
+     * @param academy           the academy to filter on
+     * @param owner             the owner to filter on
+     * @param syncTokenToIgnore the token which is considered valid, all videos with other tokens are considered
+     *                          outdated and will me marked as deleted
+     */
+    protected abstract void markOutdatedOnboardingVideosAsDeleted(String academy,
+                                                                  String owner,
+                                                                  String syncTokenToIgnore);
+
+    /**
+     * Wipes all onboarding videos for the given target/owner.
+     *
+     * @param owner the owner to delete all videos for
+     */
+    protected abstract void deleteOnboardingVideosFor(String owner);
+
+    /**
+     * Iterates over all videos of the given owner.
+     *
+     * @param owner         the onboarding participant to filter on
+     * @param videoCallback the callback to be supplied with all videos
+     */
+    protected abstract void forAllVideos(String owner, Consumer<? super OnboardingVideo> videoCallback);
+
+    /**
+     * Fetches metadata for the track (and owner).
+     *
+     * @param owner   the onboarding participant to filter on
+     * @param trackId the track to fetch metadata for
+     * @return the metadata for the given track and owner
+     */
+    public Optional<AcademyTrackInfo> fetchTrack(String owner, String trackId) {
```

Comment:
Can we move the public API of the class to the top of provided methods?

So:
- public
- protected
- private

Or are there good points that speak against this order in general? Just curious

Reply:
I commonly put a (public) API method followed by the private helper method which are used there. The class itself is layouted along its "lifecycle":
* fetch academie videos from provider
* support multiplying academy videos into onboarding videos
* query and show onboarding videos
* maintain staticstics for onboarding videos

Reply:
> a (public) API method followed by the private helper method which are used there
Yes thats perfectly fine in my eyes. But even regarding this there are protected and public Methods mixed.

The lifecycle idea is a good one but when thinking about what I am looking for when I first open a framework class is its public API as I want to get a list of all methods that I can call on the class.
The next thing could be that I want to extend the class, meaning I want a list of all abstract classes. And so on ...

But I dont want to start a big discussion about this, the most important part is that the while public API and the internal abstract API has documentation 👍